doc/memory_statistics/memory_statistics_enhancement.md

```diff
@@ -0,0 +1,338 @@
+| 1      | gNMI        | Network Management Interface, a protocol for network management and telemetry |
+| 2      | gRPC        | Remote Procedure Call, a high-performance framework for communication between network components |
+| 3      | Telemetry   | A mechanism to collect and export real-time operational data from the system |
+| 4      | Log Processing | The method of formatting and organizing log files for efficient retrieval and analysis |
+
+---
+
+## Overview
+This High-Level Design (HLD) enhances SONiC’s Memory Statistics feature, originally limited to system-level memory monitoring via CLI, into a comprehensive solution that now includes Docker containers and individual processes memory metrics. It introduces memory leak detection to prevent resource exhaustion and integrates gNMI for remote log access, reducing local dependency. This upgrade provides complete visibility across memory metrics with user-defined sampling and retention periods. Moreover, it also helps with early memory leak detection and efficient remote retrieval.
+
+---
+
+## Functional Requirements
+This section outlines the functional requirements necessary for implementing this HLD in SONiC:
+
+- **Monitoring Capabilities:** The system must monitor memory metrics for system (Total, Used, Free, Available, Cached, Buffers, Shared), Docker containers and individual processes.
+- **Memory Leak Detection:** The feature must analyze memory usage trends over time to detect potential leaks and report them via CLI.
+- **Configurability:** The system must allow configuration of sampling intervals (3–15 minutes) and retention periods (1–30 days) via CLI.
+- **CLI Enhancements:** The CLI must support displaying new metrics and leak analysis with filtering options.
+- **Remote Log Access:** gNMI integration must enable retrieval of memory logs from remote systems.
+
+---
+
+## Architecture Design
+
+The enhancement fits within the existing framework without altering its core structure. The memorystatsd is extended to collect additional metrics and detect leaks, interfacing with hostcfgd for ConfigDB updates. Enhancements were made to the existing gNMI server which processes logs into JSON and makes them remotely accessible. This integrates seamlessly with SONiC’s modular design, leveraging existing daemons and adding gNMI capabilities.
+
+<p align="center">
+    <img src="./images/architecture_diagram_v2.svg" alt="architecture diagram for memory data" width="80%"/>
+    <br>
+	Figure 1: Feature architecture diagram showing the unix socket, daemon, ConfigDB, data file and gNMI
+</p>
+
+---
+
+## High-Level Design
+
+### Core Functionalities
+
+#### Data Collection and Storage
+The `memorystatsd` collects system, Docker and process memory metrics using `psutil` and Docker APIs, storing them as compressed log files for optimized memory usage.
+
+#### Log Processing and Storage
+Logs are processed into JSON for gNMI retrieval with low overhead.
```

Comment:
Log rotation and retention need to be defined. Is logrotate being used to manage stored memory metrics?